storage/innobase/btr/btr0sea.cc

```diff
@@ -1034,1066 +1041,797 @@ TRANSACTIONAL_TARGET
-
-	if (n_fields == 0 && n_bytes == 0) {
-
-		return;
-	}
-
-	if (dict_index_get_n_unique_in_tree(index)
-	    < btr_search_get_n_fields(n_fields, n_bytes)) {
-		return;
-	}
-
-	page_t*		page	= buf_block_get_frame(block);
-	n_recs = page_get_n_recs(page);
-
-	if (n_recs == 0) {
-
-		return;
-	}
-
-	rec = page_rec_get_next_const(page_get_infimum_rec(page));
-        if (!rec) return;
-
-	if (rec_is_metadata(rec, *index)) {
-		rec = page_rec_get_next_const(rec);
-		if (!rec || !--n_recs) return;
-	}
-
-	/* Calculate and cache fold values and corresponding records into
-	an array for fast insertion to the hash index */
-
-	folds = static_cast<ulint*>(ut_malloc_nokey(n_recs * sizeof *folds));
-	recs = static_cast<const rec_t**>(
-		ut_malloc_nokey(n_recs * sizeof *recs));
-
-	n_cached = 0;
-
-	ut_a(index->id == btr_page_get_index_id(page));
+    part.latch.rd_unlock();
+fail:
+#ifdef UNIV_SEARCH_PERF_STAT
+    ++index->search_info.n_hash_fail;
+    if (index->search_info.n_hash_succ > 0)
+      --index->search_info.n_hash_succ;
+#endif /* UNIV_SEARCH_PERF_STAT */
+    index->search_info.last_hash_succ= false;
+    return false;
+  }
 
-	offsets = rec_get_offsets(
-		rec, index, offsets, index->n_core_fields,
-		btr_search_get_n_fields(n_fields, n_bytes),
-		&heap);
-	ut_ad(page_rec_is_supremum(rec)
-	      || n_fields == rec_offs_n_fields(offsets) - (n_bytes > 0));
+  const ahi_node *node= part.table.cell_get(fold)->
+    find(&ahi_node::next, [fold](const ahi_node* node)
+    { return node->fold == fold; });
 
-	fold = rec_fold(rec, offsets, n_fields, n_bytes, index->id);
+  if (!node)
+  {
+    cursor->flag= BTR_CUR_HASH_FAIL;
+    goto ahi_release_and_fail;
+  }
 
-	if (left_side) {
+  const rec_t *rec= node->rec;
+  buf_block_t *block= buf_pool.block_from_ahi(rec);
+#if defined UNIV_AHI_DEBUG || defined UNIV_DEBUG
+  ut_a(block == node->block);
+#endif
+  {
+    buf_pool_t::hash_chain &chain=
+      buf_pool.page_hash.cell_get(block->page.id().fold());
+    /* We must hold the cell latch while attempting to acquire
+    block->page.lock, because buf_LRU_block_remove_hashed() assumes
+    that block->page.can_relocate() will not cease to hold. */
+    transactional_shared_lock_guard<page_hash_latch> g
+      {buf_pool.page_hash.lock_get(chain)};
+    if (latch_mode == BTR_SEARCH_LEAF
+        ? !block->page.lock.s_lock_try()
+        : !block->page.lock.x_lock_try())
+      goto ahi_release_and_fail;
+  }
 
-		folds[n_cached] = fold;
-		recs[n_cached] = rec;
-		n_cached++;
-	}
+  /* Now that we are holding block->page.lock, the page cannot be
+  modified or evicted (buf_page_t::can_relocate() will not hold) while
+  we validate the guessed rec. Thus, releasing part.latch is safe. */
+  part.latch.rd_unlock();
 
-	while (const rec_t* next_rec = page_rec_get_next_const(rec)) {
-		if (page_rec_is_supremum(next_rec)) {
+  const auto state= block->page.state();
 
-			if (!left_side) {
+  if (UNIV_UNLIKELY(state < buf_page_t::UNFIXED))
```

Comment:
I got some further thoughts regarding this. What if instead of just reading `block->page.state()` we invoked `block->page.fix()` already here? We will be doing that a little later anyway, unless the `state` was `buf_page_t::REMOVE_HASH`. The only question is how to make sure that a possible concurrent execution of `buf_page_t::REMOVE_HASH` will not run into any trouble. A possible solution could be to protect also this increment under the `buf_pool.page_hash` latch, so that we can back off without damage if we had incremented `buf_page_t::REMOVE_HASH` to `buf_page_t::FREED`. If the `buf_page_t::lock` and the `buf_page_t::state()` field are located in different cache line, this would seem to be a bad idea: adding 1 more cache line access (a read-modify-write access at that) and conditional branch inside the critical section of a `buf_pool.page_hash` latch.